### ISSUE TITLE
GoTo Button Enhancements

### ISSUE THREAD

**Issue by iceblu3710** [MaslowCNC/GroundControl#760]
I am adding extra functionality to the `GoTo` button on screen to allow Gcode entry and Line numbers easily.

It's all working but the `_keyboard` bindings only work on numbers somehow?! I wanted the user to be able to type in `XYZ` directly and have`g` toggle `G0/G1` and `l` toggle 'L`

If I put a print in the `keydown_popup` function it is never called and on inspection `global_variables._keyboard` is always `None`. Does anybody know how the keyboard keypress bindings work in GroundControl? My Branch [HERE](https://github.com/iceblu3710/GroundControl/blob/GoTo_Changes/UIElements/touchGoToInput.py) I used the Kivy doc [HERE](https://kivy.org/docs/_modules/kivy/core/window.html) and now `_keyboard` gets an object but STILL no call into `keydown_popup`.

![groundcontrol_008](https://user-images.githubusercontent.com/187394/44558314-b860f280-a700-11e8-9bd8-ab57979e16f1.png)

**Comment by BarbourSmith** [comment]
I remember the keyboard binding being a little strange. I would look at the code which is used to bind the keyboard to the popup for entering numbers, it could be not unbinding completely causing you issues elsewhere.

I'm not 100% sure I understand the feature. The GoTo button lets you jump to a known gcode line number, when would users want to enter a line of gcode there? I totally agree that the GoTo could use a revamp because if you jump into the midst of a G2/G3 command the behavior is unpredictable

**Comment by iceblu3710** [comment]
As for why: I frequently want to jog the sled to a specific coordinate and currently I need to enter a jog value then move x/-x then a new jog value and y/-y. The GoTo popup opens with L as the default but give the easy addons of quick positioning.

Ultimately a right click/hold menu on the main canvas with a jog to here command would be ideal. This was just simpler.

I do notice the unbind call comes only then you call a new popup, not when closing the current popup. It just blows my mind that the number pad works but isn't called somehow. I will put breaks in all key calls and see if the lagging unbind is the issue. If it is that will be a separate PR for a big fix.

**Comment by iceblu3710** [comment]
Also, I have no real way of testing on a touch device. What kind of touch device/tablet/OS is the intended audience?

**Comment by iceblu3710** [comment]
Bug confirmed:
Current master branch. Open ground control and the PgUp/PgDown to zoom the canvas works. Open the GoTo or Jog value popups and upon returning to the canvas PgUp/PgDown no longer work.

**Comment by BarbourSmith** [comment]
> Fixing this binding issue will clean out A LOT of surperflous code. For EXAMPLE If the TouchNumberInput presents the user a number pad it makes sense for that widgit to handle the bind, keymapping and unbind directly. You would never override it as its a re-useable widget. That means lines 28-62 are completely unneccicary.
> 
> I wouldn't mind a second opinion on all this before I do a trial run and gut the binding in the system.

I agree with this. I've been thinking for a while that the way the touch number input is done with so much copped and pasted code there must be a better way. I think that a clean up of the system is a great idea. It will pave the way for more keyboard use.

> Ultimately a right click/hold menu on the main canvas with a jog to here command would be ideal. This was just simpler.

Is that different from the click and hold menu we have?

![image](https://user-images.githubusercontent.com/9359447/44749515-887e6a00-aac7-11e8-9eaf-cb2126ec511d.png)

**Comment by iceblu3710** [comment]
I literally had no idea this menu existed...... I have never felt the need to hold down on any one location for long periods of time....

Oh well, I found a bug making a semi-useless feature. A manual GoTo window may still be handy for some.

BTW, the hold menu displays pretty wonky on my Ubuntu machine.
![screenshot at 2018-08-28 19 58 07](https://user-images.githubusercontent.com/187394/44760607-de213980-aafc-11e8-9229-b04e62819a7d.png)
